src/librustc_mir/transform/const_prop.rs

```diff
@@ -1072,9 +1068,10 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             }
         }
         // We remove all Locals which are restricted in propagation to their containing blocks.
-        for local in self.locals_of_current_block.iter() {
-            Self::remove_const(&mut self.ecx, local);
+        for (local, mode) in self.can_const_prop.iter_enumerated() {
+            if let ConstPropMode::OnlyInsideOwnBlock = mode {
```

Comment:
Maybe it is this `if let`.

I think that in my PR, I was using at the start something rather like this loop to eliminate the `Local`s, and I had to optimize it into a `BitSet`.

Reply:
It seems, if I'm understanding the docs correctly, that the `BitIter` provided by the `BitSet` struct is faster than iterating over the `Local` list as well: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_index/bit_set/struct.BitIter.html

That + the `if let` + the `enumerate` cost might be what's causing the regression.

Reply:
Mm, indeed. The `BitSet` iteration is faster. It goes like this (I find this implementation quite incredible to be honest)

- You make a `BitIter` from the `BitSet`.
- `BitIter` uses `Word` as `u64`, and through that it calls `u64::trailing_zeros()` to compute the position of the next element in the set.
- `u64::trailing_zeros()` calls upon `intrinsics::cttz`
- And finally, `intrinsics::cttz` calls upon [the LLVM `cttz` intrinsic](https://llvm.org/docs/LangRef.html#llvm-cttz-intrinsic) which is crazy fast.

I think it's incredible because this takes the normal iteration over indexes and makes it faster. What the heck, incredible. I feel like I have so much still to learn about data structures, from looking at this. :heart:

Reply:
@oli-obk I might've figured it out.

@eddyb you know a lot about these structures and about the optimizations on them that make sense in this context. 
Do you think this regression makes sense, or are we maybe observing the change wrongly?

Reply:
Good idea @felix91gr! I've pushed a change to do that so we can try running the benchmarks again.